rareview/rareview-wordpress-coding-challenge-starter
# Add Book Listing Block to show Books created from a 'Books' Post type

# What this pull request is about
## This pull request introduces three new items
- A 'Books' Post Type
- A Book-Listing block
- A Book-Item block

These three items form the functionality to let a user add Books from the admin and then display a list of those books using the block editor. 

## The Books Post Type
The books post type has support for the title, editor, featured image (post thumbnail), excerpt and a custom post meta field field for the Author. 

Books added to the website through the Books Post Type are used by the Book Listing block to display the books.

## The Book Listing and Book Item block
The Book Listing block lets the user search for a book and display the title, author (from the custom post meta), Featured Image and excerpt in a responsive grid view. Both of these blocks are dynamic so updates made to a Book post are automatically updated where ever these blocks are used.

# Running the code
This is all built off the Rareview coding challenge starter and follows all of the conventions set out there. All you need to do is follow the instructions in [Local Development Setup](https://github.com/rareview/rareview-wordpress-coding-challenge-starter/blob/master/.lando/docs/local-development-setup.md) and you're good to go.

# How it's built
## The Books Post Type
I created a file `wp-content/themes/rareview/includes/classes/PostTypes/BookCPT.php` following the structure set up in the `ExampleCPT.php` file. The Books post type was then added to `PostTypeServiceProvider.php` to load.

One thing to note was I also added a new item to the array returned by the `label()` function in `PostType.php` called `add_new`. This is because WP version 6.4 introduced a change where `add_new_item` doesn't affect the label in the sidebar of the admin but `add_new` does. 

### The author custom post meta
Adding the author name meta field was done in a similar way where I followed an existing structure. In this case I created a new file `wp-content/themes/rareview/assets/js/admin/gutenberg/custom-sidebars/book/BookCptAuthorNameMeta.js` based off `ExampleCptExampleMeta.js`. This file is almost exactly the same but it updates the `_rareview_book_author_name` post meta.

I did find a bug where the `<MetaTextControlInput>` component was exporting as `TextControlPickerInput`. I think this is a bug as `ExampleCptExampleMeta.js` was also importing the component as `MetaTextControlInput`. I changed the name of the component to fix this.

## The Book Listing and Book Item blocks
These blocks are built to be used together. Book Item can only be used within Book Listing.

### Book Listing Block
The book listing block uses the `<ContentSearch>` component from `@10/block-components` as it gives us some nice searching and adding capabilities built in. The search is limited to only 'books' post type. It also uses `useInnerBlockProps` to show and reorganise books via drag and drop.

On selecting a book `insertBook()` is called. This function creates a book item block, passing through the post id of the selected book, it then fires a dispatch from the WP data api to insert the newly created block.

The InnerBlocks in the Book Listing block is restricted to only allow book item blocks in it. This prevents the user from adding unwanted blocks inside.

### Book Item block
This block is specifically made to only be usable inside the book listing block. This is set in the `parent` property in block.json. It is passed an id which is used by `getEntityRecord` (via the useSelect hook) to get the post and populate the details for the view in the editor. I have also split out the FeaturedImage into it's own component as it also has a useSelect hook to get the image URL. It is a dynamic block so the frontend is generated by markup.php and used the post id saved in its 'id' attribute.

In terms of markup, these blocks are created as a `<ul>` with book items being a `<li>`. I felt this was appropriate from a semantics point of view as we are display a **list** of books. I have also included the alt tag for the featured image to further improve accessibility.

### Restricting usage of the book listing block. 
The book listing block can only be used on 'Posts'. 
In the `wp-content/themes/rareview/includes/blocks.php` file. I have hooked into the `set_allowed_block_types` filter. Because this works off an allow list approach (i.e you return the blocks that you want to allow). I also had to use the `WP_Block_Type_Registry` class as it contains a method to `get_all_registered()` blocks. From there I checked the post type and removed the book listing block from that array if the post type was not 'post.

# How to use and test
1. You need to create some books to test the blocks
    - Go to the books post type
    - Click on 'Add new Book'
    - Enter the title, author name, featured image and excerpt.
    - Hit Publish
2. Now you can go and create a new post and insert the block
    - Type `/books listing` and the Book Listing block should show up.
    - Once inserted you can search for the book in the search bar.
    - After adding your books you can drag and drop to rearrange them as you would any InnerBlock children.

## On testing
Currently the only way to test is through manual testing. But I feel some automated testing would be useful here. I would love to get some feedback on how that could be implemented.

# Misc
I also added 'Inter' for a bit of fun (and it's my favourite sans-serif font).
